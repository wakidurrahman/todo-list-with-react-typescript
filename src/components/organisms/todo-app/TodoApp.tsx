/**
 * TodoApp Component
 *
 * A React component that implements a complete todo list application with CRUD operations.
 * Features include:
 * - Viewing all todos with filtering and sorting capabilities
 * - Adding new todos with validation
 * - Editing existing todos with history tracking
 * - Deleting todos with confirmation
 * - Marking todos as complete/incomplete
 * - Persisting todos in localStorage or remote API
 * - Responsive layout with Bootstrap
 * - Accessibility features
 * - Error handling and loading states
 *
 * @component
 * @example
 * ```tsx
 * <TodoApp />
 * ```
 */

import { useEffect, useState } from 'react';

import useError from '../../../hooks/useError';
import { Todo, TodoFormData } from '../../../types/todo.types';
import {
  addTodo,
  deleteTodo,
  getTodos,
  updateTodo,
} from '../../../utils/storage';
import Alert from '../../atoms/alert/Alert';
import TodoForm from '../../molecules/todo-form/TodoForm';
import TodoList from '../../molecules/todo-list/TodoList';

const TodoApp = () => {
  const [loading, setLoading] = useState(true);
  const [todos, setTodos] = useState<Todo[]>([]);
  const [todoBeingEdited, setTodoBeingEdited] = useState<Todo | null>(null);
  const { error, setError, clearError } = useError();

  useEffect(() => {
    const handleLoadTodos = async () => {
      try {
        const todos = await getTodos();
        setTodos(todos);
      } catch (error) {
        setLoading(false);
        setError('Failed to load todos. Please try again later.', 'danger');
      } finally {
        setLoading(false);
      }
    };
    handleLoadTodos();
  }, [setError]);

  const handleAddTodo = async (data: TodoFormData) => {
    try {
      const newTodo = await addTodo(data);
      setTodos((prev) => [...prev, newTodo]);
      setError('Todo added successfully!', 'primary', 3000);
    } catch (error) {
      setError('Failed to add todo. Please try again.', 'danger');
    }
  };

  const handleUpdateTodo = async (data: TodoFormData) => {
    if (!todoBeingEdited) return;

    try {
      const updatedTodo = await updateTodo(todoBeingEdited.id, data);
      if (!updatedTodo) {
        setError('Todo not found or already deleted.', 'warning');
        return;
      }

      setTodos((prev) =>
        prev.map((todo) =>
          todo.id === todoBeingEdited.id ? updatedTodo : todo
        )
      );
      setTodoBeingEdited(null);
      setError('Todo updated successfully!', 'secondary', 3000);
    } catch (error) {
      setError('Failed to update todo. Please try again.', 'danger');
    }
  };

  const handleDeleteTodo = async (id: string) => {
    try {
      const isDeleted = await deleteTodo(id);
      if (!isDeleted) {
        setError('Todo not found or already deleted.', 'warning');
        return;
      }

      setTodos((prev) => prev.filter((todo) => todo.id !== id));
      setError('Todo deleted successfully!', 'danger', 3000);
    } catch (error) {
      setError('Failed to delete todo. Please try again.', 'danger');
    }
  };

  const handleEditClick = (todo: Todo) => {
    setTodoBeingEdited(todo);
  };

  const handleCancelEdit = () => {
    setTodoBeingEdited(null);
  };

  const handleToggleComplete = async (id: string) => {
    const todo = todos.find((item) => item.id === id);
    if (!todo) {
      setError('Todo not found.', 'warning');
      return;
    }

    try {
      const updatedTodo = await updateTodo(id, {
        ...todo,
        completed: !todo.completed,
      });
      if (!updatedTodo) {
        setError('Failed to update todo status.', 'warning');
        return;
      }

      setTodos((prev) =>
        prev.map((item) => (item.id === id ? updatedTodo : item))
      );
      setError(
        `Todo marked as ${updatedTodo.completed ? 'completed' : 'incomplete'}!`,
        'success',
        3000
      );
    } catch (error) {
      setError('Failed to update todo status. Please try again.', 'danger');
    }
  };

  return (
    <div className="container py-4">
      {error.show && (
        <Alert
          message={error.message}
          variant={error.variant}
          show={error.show}
          onClose={clearError}
          timeout={error.timeout}
        />
      )}
      <h1 className="text-center mb-4">Todo List Application</h1>
      <div className="row">
        <div className="col-12 col-sm-5 col-md-6">
          <div className="card p-3 mb-4">
            <h3 className="mb-3">
              {todoBeingEdited ? 'Edit Todo' : 'Add New Todo'}
            </h3>
            <TodoForm
              onSubmit={todoBeingEdited ? handleUpdateTodo : handleAddTodo}
              editingTodo={todoBeingEdited || undefined}
              onCancel={handleCancelEdit}
            />
          </div>
        </div>
        <div className="col-12 col-sm-7 col-md-6">
          <TodoList
            loading={loading}
            todos={todos}
            onEdit={handleEditClick}
            onDelete={handleDeleteTodo}
            onToggleComplete={handleToggleComplete}
          />
        </div>
      </div>
    </div>
  );
};

export default TodoApp;
