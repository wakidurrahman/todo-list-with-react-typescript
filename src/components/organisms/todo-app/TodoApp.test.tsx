import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import TodoApp from './TodoApp';

// Mock localStorage utility functions
jest.mock('../../../utils/storage', () => ({
  getTodos: jest.fn(),
  addTodo: jest.fn(),
  updateTodoInAPI: jest.fn(),
  deleteTodoInAPI: jest.fn(),
}));

// Add this line after the mock definition
const { getTodos, addTodo, updateTodoInAPI, deleteTodoInAPI } =
  jest.requireMock('../../../utils/storage');

describe('TodoApp Component', () => {
  const mockTodos = [
    {
      id: '1',
      title: 'First Todo',
      description: 'First Description',
      completed: false,
      createdAt: new Date('2024-02-16T00:00:00.000Z'),
      updatedAt: new Date('2024-02-16T00:00:00.000Z'),
    },
    {
      id: '2',
      title: 'Second Todo',
      description: 'Second Description',
      completed: false,
      createdAt: new Date('2024-02-16T01:00:00.000Z'),
      updatedAt: new Date('2024-02-16T01:00:00.000Z'),
    },
  ];

  const mockOnError = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    (getTodos as jest.Mock).mockResolvedValue(mockTodos);
    (addTodo as jest.Mock).mockImplementation((data) => ({
      id: '3',
      ...data,
      completed: false,
      createdAt: new Date(),
      updatedAt: new Date(),
    }));
    (updateTodoInAPI as jest.Mock).mockImplementation((id, data) => ({
      id,
      ...data,
      createdAt: new Date(),
      updatedAt: new Date(),
    }));
    (deleteTodoInAPI as jest.Mock).mockResolvedValue(true);
  });

  // Snapshot Tests
  describe('Snapshots', () => {
    it('renders correctly with todos', () => {
      const { container } = render(<TodoApp onError={mockOnError} />);
      expect(container).toMatchSnapshot();
    });

    it('renders correctly without todos', () => {
      (getTodos as jest.Mock).mockResolvedValue([]);
      const { container } = render(<TodoApp onError={mockOnError} />);
      expect(container).toMatchSnapshot();
    });
  });

  // Unit Tests
  describe('Functionality', () => {
    const user = userEvent.setup();

    it('loads todos and shows error on failure', async () => {
      const error = new Error('Failed to load todos');
      (getTodos as jest.Mock).mockRejectedValueOnce(error);

      render(<TodoApp onError={mockOnError} />);

      await waitFor(() => {
        expect(mockOnError).toHaveBeenCalledWith(
          'Failed to load todos. Please try again later.',
          'danger'
        );
      });
    });

    it('adds a new todo successfully', async () => {
      render(<TodoApp onError={mockOnError} />);

      const titleInput = screen.getByLabelText(/title/i);
      const descriptionInput = screen.getByLabelText(/description/i);
      await user.type(titleInput, 'New Todo');
      await user.type(descriptionInput, 'New Todo Description');

      const addButton = screen.getByRole('button', { name: /add/i });
      await user.click(addButton);

      expect(addTodo).toHaveBeenCalledWith({
        title: 'New Todo',
        description: 'New Todo Description',
      });

      await waitFor(() => {
        expect(mockOnError).toHaveBeenCalledWith(
          'Todo added successfully!',
          'success',
          3000
        );
      });
    });

    it('shows error when adding todo fails', async () => {
      (addTodo as jest.Mock).mockRejectedValueOnce(new Error('Failed to add'));
      render(<TodoApp onError={mockOnError} />);

      const titleInput = screen.getByLabelText(/title/i);
      const descriptionInput = screen.getByLabelText(/description/i);
      await user.type(titleInput, 'New Todo');
      await user.type(descriptionInput, 'New Todo Description');

      const addButton = screen.getByRole('button', { name: /add/i });
      await user.click(addButton);

      await waitFor(() => {
        expect(mockOnError).toHaveBeenCalledWith(
          'Failed to add todo. Please try again.',
          'danger'
        );
      });
    });

    it('deletes a todo successfully', async () => {
      render(<TodoApp onError={mockOnError} />);

      await waitFor(() => {
        expect(screen.getByText('First Todo')).toBeInTheDocument();
      });

      const deleteButton = screen.getAllByRole('button', {
        name: /delete/i,
      })[0];
      await user.click(deleteButton);

      expect(deleteTodoInAPI).toHaveBeenCalledWith('1');

      await waitFor(() => {
        expect(mockOnError).toHaveBeenCalledWith(
          'Todo deleted successfully!',
          'success',
          3000
        );
      });
    });

    it('shows error when todo deletion fails', async () => {
      (deleteTodoInAPI as jest.Mock).mockRejectedValueOnce(
        new Error('Failed to delete')
      );
      render(<TodoApp onError={mockOnError} />);

      await waitFor(() => {
        expect(screen.getByText('First Todo')).toBeInTheDocument();
      });

      const deleteButton = screen.getAllByRole('button', {
        name: /delete/i,
      })[0];
      await user.click(deleteButton);

      await waitFor(() => {
        expect(mockOnError).toHaveBeenCalledWith(
          'Failed to delete todo. Please try again.',
          'danger'
        );
      });
    });

    it('shows warning when deleting non-existent todo', async () => {
      (deleteTodoInAPI as jest.Mock).mockResolvedValueOnce(false);
      render(<TodoApp onError={mockOnError} />);

      await waitFor(() => {
        expect(screen.getByText('First Todo')).toBeInTheDocument();
      });

      const deleteButton = screen.getAllByRole('button', {
        name: /delete/i,
      })[0];
      await user.click(deleteButton);

      await waitFor(() => {
        expect(mockOnError).toHaveBeenCalledWith(
          'Todo not found or already deleted.',
          'warning'
        );
      });
    });

    it('updates todo completion status successfully', async () => {
      render(<TodoApp onError={mockOnError} />);

      await waitFor(() => {
        expect(screen.getByText('First Todo')).toBeInTheDocument();
      });

      const checkbox = screen.getByLabelText('First Todo');
      await user.click(checkbox);

      expect(updateTodoInAPI).toHaveBeenCalledWith('1', {
        ...mockTodos[0],
        completed: true,
      });

      await waitFor(() => {
        expect(mockOnError).toHaveBeenCalledWith(
          'Todo marked as completed!',
          'success',
          3000
        );
      });
    });

    it('shows error when updating todo completion fails', async () => {
      (updateTodoInAPI as jest.Mock).mockRejectedValueOnce(
        new Error('Failed to update')
      );
      render(<TodoApp onError={mockOnError} />);

      await waitFor(() => {
        expect(screen.getByText('First Todo')).toBeInTheDocument();
      });

      const checkbox = screen.getByLabelText('First Todo');
      await user.click(checkbox);

      await waitFor(() => {
        expect(mockOnError).toHaveBeenCalledWith(
          'Failed to update todo status. Please try again.',
          'danger'
        );
      });
    });
  });
});
