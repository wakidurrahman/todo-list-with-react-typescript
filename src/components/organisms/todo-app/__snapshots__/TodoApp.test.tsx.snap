// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoApp Component Snapshots renders correctly with todos 1`] = `
<div>
  <div
    class="container py-4"
  >
    <h1
      class="text-center mb-4"
    >
      Todo List Application
    </h1>
    <div
      class="row"
    >
      <div
        class="col-12 col-sm-5 col-md-6"
      >
        <div
          class="card p-3 mb-4"
        >
          <h3
            class="mb-3"
          >
            Add New Todo
          </h3>
          <form
            class="row g-3"
            novalidate=""
          >
            <div
              class="col-12"
            >
              <label
                class="form-label"
                for="title"
              >
                Title
              </label>
              <input
                aria-describedby="titleFeedback"
                aria-label="Todo title"
                class="form-control "
                id="title"
                maxlength="50"
                name="title"
                placeholder="Enter title"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
            <div
              class="col-12"
            >
              <label
                class="form-label"
                for="description"
              >
                Description
              </label>
              <textarea
                aria-describedby="descriptionFeedback"
                aria-label="Todo description"
                class="form-control "
                id="description"
                maxlength="500"
                name="description"
                placeholder="Enter description"
                rows="3"
                tabindex="0"
              />
            </div>
            <div
              class="col-12 d-flex justify-content-center gap-2"
            >
              <button
                aria-label="Add todo"
                class="btn btn-primary btn-md"
                type="submit"
              >
                Add
                 Todo
              </button>
              <button
                aria-label="Cancel editing"
                class="btn btn-secondary btn-md"
                type="button"
              >
                Cancel
              </button>
            </div>
          </form>
        </div>
      </div>
      <div
        class="col-12 col-sm-7 col-md-6"
      >
        <div
          class="d-flex justify-content-center align-items-center gap-3 flex-wrap"
        >
          <div
            class="spinner-border text-primary  "
            role="status"
          >
            <span
              class="visually-hidden"
            >
              Loading...
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TodoApp Component Snapshots renders correctly without todos 1`] = `
<div>
  <div
    class="container py-4"
  >
    <h1
      class="text-center mb-4"
    >
      Todo List Application
    </h1>
    <div
      class="row"
    >
      <div
        class="col-12 col-sm-5 col-md-6"
      >
        <div
          class="card p-3 mb-4"
        >
          <h3
            class="mb-3"
          >
            Add New Todo
          </h3>
          <form
            class="row g-3"
            novalidate=""
          >
            <div
              class="col-12"
            >
              <label
                class="form-label"
                for="title"
              >
                Title
              </label>
              <input
                aria-describedby="titleFeedback"
                aria-label="Todo title"
                class="form-control "
                id="title"
                maxlength="50"
                name="title"
                placeholder="Enter title"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
            <div
              class="col-12"
            >
              <label
                class="form-label"
                for="description"
              >
                Description
              </label>
              <textarea
                aria-describedby="descriptionFeedback"
                aria-label="Todo description"
                class="form-control "
                id="description"
                maxlength="500"
                name="description"
                placeholder="Enter description"
                rows="3"
                tabindex="0"
              />
            </div>
            <div
              class="col-12 d-flex justify-content-center gap-2"
            >
              <button
                aria-label="Add todo"
                class="btn btn-primary btn-md"
                type="submit"
              >
                Add
                 Todo
              </button>
              <button
                aria-label="Cancel editing"
                class="btn btn-secondary btn-md"
                type="button"
              >
                Cancel
              </button>
            </div>
          </form>
        </div>
      </div>
      <div
        class="col-12 col-sm-7 col-md-6"
      >
        <div
          class="d-flex justify-content-center align-items-center gap-3 flex-wrap"
        >
          <div
            class="spinner-border text-primary  "
            role="status"
          >
            <span
              class="visually-hidden"
            >
              Loading...
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
